staging/src/k8s.io/kubectl/pkg/cmd/apply/applyset.go

```diff
@@ -141,10 +143,24 @@ func NewApplySet(parent *ApplySetParentRef, tooling ApplySetTooling, mapper meta
 	}
 }
 
+const applySetIDPartDelimiter = "."
+
 // ID is the label value that we are using to identify this applyset.
+// Format: base64(sha256(<name>.<namespace>.<kind>.<group>)), using the URL safe encoding of RFC4648.
 func (a ApplySet) ID() string {
-	// TODO: base64(sha256(gknn))
-	return "placeholder-todo"
+	var unencoded string
+	if a.parentRef.IsNamespaced() {
+		unencoded = strings.Join([]string{a.parentRef.Name, a.parentRef.Namespace, a.parentRef.GroupVersionKind.Kind, a.parentRef.GroupVersionKind.Group}, applySetIDPartDelimiter)
+	} else {
+		unencoded = strings.Join([]string{a.parentRef.Name, a.parentRef.GroupVersionKind.Kind, a.parentRef.GroupVersionKind.Group}, applySetIDPartDelimiter)
```

Comment:
Is this what we want for non-namespaced? The omission of a segment + use of a delimiter that appears inside the last segment could make correct parsing of the value difficult... but AFAIK tooling should only ever reconstruct this for verification, not attempt to parse it. Wdyt?

Reply:
Good question.  I'd argue for including the empty namespace (i.e. be consistent even if less aesthetically pleasing), because no human is going to notice the empty value once it goes through sha256/base64, and if we ever do need to parse it, then the ambiguous form is going to cause us trouble.

I think we might need to / want to parse it in our `fsck` command.  But otherwise, yes I agree it should not be parsed.

This is not a strongly held opinion - if you want to go the other way that works too!